### PR TITLE
Shard awareness extension for Scylla

### PR DESCRIPTION
**Pull request by dahankzter**
This PR brings shard awareness to the token aware policy for Scylla servers version 2.3 or later.

This is done by implementing the Scylla protocol extension mechanism as as defined by: 
https://github.com/scylladb/scylla/blob/master/docs/protocol-extensions.md
The impact for Scylla users is significant improvement in performance while still maintaining the original performance for Cassandra users.
Both latency and throughput are greatly improved but especially latency can see an order of magnitude improvement.

No config is needed for the user to benefit from this patch, the driver will discover the sharding support on its own.

### CODE REVIEW

File: conn_test.go [hunk 1]
```diff
@@ -828,14 +828,19 @@ func TestFrameHeaderObserver(t *testing.T) {
 
 	frames := observer.getFrames()
 
-	if len(frames) != 2 {
+	if len(frames) != 3 {
 		t.Fatalf("Expected to receive 2 frames, instead received %d", len(frames))
```
[the remark this answers is not in the document]
Comment (dahankzter) [reply conn_test.go]: Ouch @JvGinkel I read your message too loosely it seems.
I will correct the message. Good catch!

File: connpicker.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	Put(*Conn)
+	Remove(conn *Conn)
+	Size() (int, int)
+	Close()
+}
+
+type DefaultConnPicker struct {
+	conns []*Conn
+	pos   uint32
+	size  int
```
Comment (alourie) [inline connpicker.go]: I'd use an unsigned int variant for size.

File: connpicker.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	mu    sync.RWMutex
+}
+
+func NewDefaultConnPicker(size int) *DefaultConnPicker {
+	return &DefaultConnPicker{
```
Comment (alourie) [inline connpicker.go]: check for size to be non-zero to avoid errors?

File: connpicker.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	}
+}
+
+func (p *DefaultConnPicker) Remove(conn *Conn) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for i, candidate := range p.conns {
+		if candidate == conn {
+			p.conns[i] = conn
```
Comment (alourie) [inline connpicker.go]: `p.conns[i] == nil`?

File: connpicker.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	}
+}
+
+func (p *DefaultConnPicker) Size() (int, int) {
+	sz := len(p.conns)
+	return sz, p.size - sz
+}
+
+func (p *DefaultConnPicker) Pick(token) *Conn {
+	pos := int(atomic.AddUint32(&p.pos, 1) - 1)
```
Comment (alourie) [inline connpicker.go]: Not sure the initial reason for this, but would you mind just reading atomically directly?

File: policies.go [hunk 6]
```diff
@@ -305,16 +305,24 @@ type HostSelectionPolicy interface {
...
 }
 
-type selectedHost HostInfo
+type selectedHost struct {
+	info  *HostInfo
+	token token
+}
+
+func (host selectedHost) Info() *HostInfo {
+	return (*HostInfo)(host.info)
```
Comment (alourie) [inline policies.go]: probably no need to cast here

File: connpicker.go (hunk 5 above)
Comment (dahankzter) [reply connpicker.go]: This keeps track of the current position in the conns list. I think it was always like this to step to the next connection in one atomic go.

**Comment by dahankzter** [comment]
About your major comment @alourie I would also want to isolate and modularize it more but when we implemented this we saw now real option without major refactoring which felt very invasive for such a relatively small patch. It depends of course on what you mean with this backends module? A factory struct that knows how to instantiate backends? 

I will try to address your minor issues asap, thanks for noticing. Nitpicks accumulate so it's good to catch them early!

File: connpicker.go (hunk 2 above)
Comment (dahankzter) [reply connpicker.go]: It's very viral. This property is threaded through the code and it makes sense to have them as `uint` but perhaps not in this patch?

File: connpicker.go (hunk 3 above)
Comment (dahankzter) [reply connpicker.go]: I can check it here and return an error. If so what do we do then, panic or default to something?

Comment (dahankzter) [reply connpicker.go]: There is no real infrastructure available to handle errors in this. It all relies on the config defaulting to 2. We can change the assumptions but I'd rather not in a patch that is about something else.

What do you think?

File: connpicker.go (hunk 4 above)
Comment (dahankzter) [reply connpicker.go]: `p.conns[i] = nil` perhaps? I think we should maybe eve resize it what do you think?

Comment (dahankzter) [reply connpicker.go]: I have set it to nil now. We can change it if you prefer but the rest of the code works guards against nil.

File: policies.go (hunk 6 above)
Comment (dahankzter) [reply policies.go]: Ah, thx!

File: connpicker.go (hunk 3 above)
Comment (alourie) [reply connpicker.go]: I think panic here is fine.

File: connpicker.go (hunk 5 above)
Comment (alourie) [reply connpicker.go]: right, ok

File: connpicker.go (hunk 4 above)
Comment (alourie) [reply connpicker.go]: nah, nil is fine. That's exactly what I meant :-)

**Comment by alourie** [comment]
@dahankzter I don't know what's the best way to implement backends, for now, would need to wait until @Zariel has an opinion. Maybe in the next major API change.

But splitting this into 2 PRs should not be complicated, no? The connection picker and the Scylla's implementation are quite separate, so I'd assume it would be possible to do.

**Comment by dahankzter** [comment]
We can probably split them up but the assignment of the picker to use is dynamic so at some point the code will blend.

There are basically 2 reasons for the code being as is today.

1. Make as small an impact as possible on the core code.
2. Make the usage for users as small as possible. They only need to update the driver now for it to kick in.

That's it really we can make any sort of bootstrapping otherwise if we don't want 2.
Maybe there even is a way to implement it more isolated that we haven't thought of and while we prefer no-config for end users we are not married to the idea. We want to be pragmatic about it and a config option can work fine as well. We just need to tell users to enable it.

**Comment by mmatczuk** [comment]
Hi @alourie,
I have fixed the code to limit scope of Scylla related stuff as much possible. It's split into multiple commits and two PRs. Please find the gocql generic changes in #1210 and Scylla specific changes in #1211 (note the dependency).

cc @dahankzter

**Comment by dahankzter** [comment]
Lets work off of #1210 and #1211 instead.

**Comment by alourie** [comment]
:+1:

**Pull request closed without merging**